From ab76a6f63e3a95c0c787d920b143df4c2464cd5e Mon Sep 17 00:00:00 2001
From: Lucas Alvares Gomes <lucasagomes@gmail.com>
Date: Fri, 9 Jan 2015 14:36:32 +0000
Subject: [PATCH] Ironic: Local boot

This patch extends the deploy-ironic element to make it install a
bootloader on the disk in case the "localboot" flag is passed via the
kernel cmdline.

Change-Id: I5ebe6f364ae0ac408939399e5f28728b41c4766e

Conflicts:
	elements/deploy-ironic/binary-deps.d/deploy-ironic
	elements/deploy-ironic/package-installs.yaml
---
 elements/deploy-ironic/binary-deps.d/deploy-ironic |  1 +
 elements/deploy-ironic/init.d/80-deploy-ironic     | 53 ++++++++++++++++++++++
 .../install.d/package-installs-deploy-ironic       |  1 +
 3 files changed, 55 insertions(+)

diff --git a/elements/deploy-ironic/binary-deps.d/deploy-ironic b/elements/deploy-ironic/binary-deps.d/deploy-ironic
index 13368f8..6db325a 100644
--- a/elements/deploy-ironic/binary-deps.d/deploy-ironic
+++ b/elements/deploy-ironic/binary-deps.d/deploy-ironic
@@ -1 +1,2 @@
 curl
+partprobe
diff --git a/elements/deploy-ironic/init.d/80-deploy-ironic b/elements/deploy-ironic/init.d/80-deploy-ironic
index d52ccde..ed4c116 100644
--- a/elements/deploy-ironic/init.d/80-deploy-ironic
+++ b/elements/deploy-ironic/init.d/80-deploy-ironic
@@ -1,4 +1,5 @@
 readonly IRONIC_API_URL=$(get_kernel_parameter ironic_api_url)
+readonly IRONIC_LOCALBOOT=$(get_kernel_parameter localboot)
 
 if [ -z "$ISCSI_TARGET_IQN" ]; then
   err_msg "iscsi_target_iqn is not defined"
@@ -59,3 +60,55 @@ nc -l -p 10000
 echo "stop iSCSI target on $target_disk"
 
 stop_iscsi_target
+
+# If localboot is set, install a bootloader
+if [ "$IRONIC_LOCALBOOT" ]; then
+    echo "Installing bootloader"
+
+    # We need to run partprobe to ensure all partitions are visible
+    partprobe $target_disk
+
+    # root partition is always the last partition of the disk
+    readonly root_part=$(ls $target_disk* | tr " " "\n" | tail -n1)
+    readonly root_part_mount=/mnt/rootfs
+
+    mkdir -p $root_part_mount
+
+    mount $root_part $root_part_mount
+    mount -o bind /dev $root_part_mount/dev
+    mount -o bind /sys $root_part_mount/sys
+    mount -o bind /proc $root_part_mount/proc
+
+    # TODO(lucasagomes): Add extlinux as a fallback
+    # Find grub version
+    V=
+    if [ -x $root_part_mount/usr/sbin/grub2-install ]; then
+        V=2
+    fi
+
+    # Install grub
+    ret=1
+    if chroot $root_part_mount /bin/bash -c "/usr/sbin/grub$V-install ${target_disk}"; then
+        echo "Generating the grub configuration file"
+
+        # tell GRUB2 to preload its "lvm" module to gain LVM booting on direct-attached disks
+        if [ "$V" = "2" ]; then
+            echo "GRUB_PRELOAD_MODULES=lvm" >> $root_part_mount/etc/default/grub
+        fi
+
+        chroot $root_part_mount /bin/bash -c "/usr/sbin/grub$V-mkconfig -o /boot/grub$V/grub.cfg"
+        ret=$?
+    fi
+
+    umount $root_part_mount/dev
+    umount $root_part_mount/sys
+    umount $root_part_mount/proc
+    umount $root_part_mount
+
+    if [ $ret -eq 0 ]; then
+         echo "Bootloader successfully installed"
+    fi
+
+    # TODO(lucasagomes): This is just for the PoC, remove it later
+    poweroff -f
+fi
diff --git a/elements/deploy-ironic/install.d/package-installs-deploy-ironic b/elements/deploy-ironic/install.d/package-installs-deploy-ironic
index 3da7a78..5ba802e 100644
--- a/elements/deploy-ironic/install.d/package-installs-deploy-ironic
+++ b/elements/deploy-ironic/install.d/package-installs-deploy-ironic
@@ -1,2 +1,3 @@
 curl
 tgt
+parted
-- 
1.9.3

